src/daft-io/src/s3_like.rs

```diff
@@ -952,6 +1011,30 @@ impl ObjectSource for S3LikeSource {
         }
     }
 
+    async fn put(
+        &self,
+        uri: &str,
+        data: bytes::Bytes,
+        io_stats: Option<IOStatsRef>,
+    ) -> super::Result<()> {
+        let data_len = data.len();
+        let permit = self
+            .connection_pool_sema
+            .clone()
+            .acquire_owned()
+            .await
+            .context(UnableToGrabSemaphoreSnafu)?;
+        self._put_impl(permit, uri, data, &self.default_region)
+            .await?;
```

Comment:
If we do `?` here, what happens to the errors from `_put_impl`?

Reply:
Short circuits, returns the `Err` immediately!